### ISSUE TITLE
Intermittent Stripe alternative payment failures (iDeal / Multibanco / SOFORT)

### ISSUE THREAD

**Issue by grig-23** [woocommerce/woocommerce-gateway-stripe#1196]
**Describe the bug**
Multiple users report these intermittent failures when payments are done via Stripe iDeal:

1. After a source is created, in order for the source object to become chargeable the customer has to authorize the transfer of the funds during the redirect to the iDeal site.
2. When the customer authorizes the charge, the funds are deducted from the customer's bank account immediately. At this point the Source object becomes chargeable and must be used within six hours.
3. The source is not added to order's billing details or charged within that time. User in 2950247-zen found it in the logs though. Its status is automatically transitioned to canceled in Stripe and our extension receives a source.canceled webhook.
4. Stripe issues a reversal of the payment within 2-3 days to the customer's bank account.
5. When a source is canceled, the customer’s authenticated payment is refunded automatically—no money is moved into merchant's account. The customer will see the money move out of their account and then back into their account after 3-5 days depending on their bank processes. 
6. Canceled sources can be viewed in mechant's Dashboard by searching for status:canceled, see https://dashboard.stripe.com/search?query=status%3Acanceled .

**To Reproduce**
We were not able to reproduce this behaviour.

The response shows `null` next to the bank details in the logs copied from Stripe dashboard in 2889513-zen.

In one of the tickets, there are some errors in the error log on Woo side but those are most likely unrelated since we don't see anything like that in other tickets.

Multiple users report it though: 

2872699-zen
2924615-zen
2950247-zen
2889513-zen
3021386-zen

There's also a ticket where the same thing is happening with Multibanco payments: 2948254-zen

**Expected behavior**
If the funds are deducted from user's bank account - the source should be added to the billing details in WooCommerce and charged in Stripe. That doesn't always happen.

**Comment by FreshPhil** [comment]
As @grigorijurasov mentioned, it seems to be more than just iDeal.

If seen multiple tickets following this trend and they are all for off-site methods (iDeal, Multibanco, Bancontact).

Examples:

2970134-zen (iDeal)
2954013-zen (iDeal)
2971790-zen (Multibanco)
2957800-zen (Bancontact)

**Comment by treibalen** [comment]
Can we bump it to Critical @allendav? There are way too many similar reports of the issue

**Comment by iMazed** [comment]
Another report here: 2976620-zen. I also vote to set this to Critical!

**Comment by FreshPhil** [comment]
Stripe support went through a full iDeal test with a mutual customer and they shared that exchange with us in 2967914-zen. Please see their response from May 19th for that conversation.

TLDR: Stripe said that sources were created, however WooCommerce didn't send a request to charge the source so it expired after not being captured.

**Comment by dechov** [comment]
Did some investigation on this today, but still don't have a clear picture of what the cause is likely to be, and whether this is a single issue or multiple.

Looked through some of the relevant code but haven't seen any potential edge case or race condition that seems likely to cause a prevalent issue.

Charging the source some or most of the time would be consistent with the webhook just not being configured or not working at all, since it does charge on redirect as long as the source status has already switched to `chargeable` at that point.

It would help to know (via Stripe) whether the webhook was attempted and/or successfully received by the site (for a source that was chargeable before being canceled), though Stripe's Dashboard only keeps this information for 15 days so it might be tricky to get this confirmed.

**Comment by yukikatayama** [comment]
@dechov 

> It would help to know (via Stripe) whether the webhook was attempted and/or successfully received by the site (for a source that was chargeable before being canceled), though Stripe's Dashboard only keeps this information for 15 days so it might be tricky to get this confirmed.

To confirm, we should ask users for any logs found under Developers > Webhooks in Stripe, or the logs from the order itself?

**Comment by dechov** [comment]
@yukikatayama Yeah, in the Stripe Dashboard under Developers »  Events, you can click into an event and see if it was attempted, delivered, and what the response was. The case we're after is when a source is chargeable and then canceled (since it can also be canceled without ever having been chargeable, which is not the issue here).

Not sure how to search directly for a source like this, but one way would be to look through https://dashboard.stripe.com/events?type=source.canceled, click on an event, click on the ID (`src_xxxxxxx`) in the Event data, scroll to Events at bottom and check if there is an "is chargeable" event below the "was canceled" one. If so, click it and see what's in the "Webhook attempts" table (though it'll be empty in any case if the event was over 15 days ago). If not, go back and check a different "canceled" source.

---

Just a side-note on the webhook: it's possible that it's configured properly but failing due to the system time being more than 5 minutes off (causing it to fail [this check](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/19e5f4a47d4ad94d4d367a588b3cd4b227c9f80d/includes/class-wc-stripe-webhook-handler.php#L106)). The solution would be to change the system time, or otherwise to just unset the webhook secret in the gateway settings, in which case the webhook events wouldn't be verified in this way.

**Comment by damiannep** [comment]
Another case of this in 2970134-zen

**Comment by damiannep** [comment]
Another case in 2996522-zen. Reported to Stripe as well.

**Comment by FreshPhil** [comment]
Another case for the same issue with offsite payments (Multibanco) 3020725-zen

I've asked them to look in their Stripe account for the events and webhook info as per the comment above.

**Comment by dechov** [comment]
If the webhook is not configured and enabled – or is failing verification, which will happen if the server time is more than 5 minutes off for instance – then the source will not be charged in _some_ cases (and will be in others, since the redirect also triggers charge of the source but _only_ if the source has become chargeable soon enough – it's a race condition). It seems likely that this is responsible for some if not most of the cases reported, though it's hard to be sure.

There is some work underway to expose more webhook-related messaging to the merchant (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1224, https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1226), and I'm hopeful this will mitigate this issue to some extent, but still welcome any reports of chargeable sources not being charged, especially (with regard to this particular issue) if there was a (failed or successful) webhook attempt on the event.

**Comment by stephhjacq** [comment]
Hi @dechov, please see this user's logs as requested, showing chargeable source failing, from 3021386-zen

Overview of events:
![https://d.pr/i/ccTg2n](https://d.pr/i/ccTg2n+)
Full Image: https://d.pr/i/ccTg2n

Event "...is chargeable":
![https://d.pr/i/IYLEKC](https://d.pr/i/IYLEKC+)
Full Image: https://d.pr/i/IYLEKC

Webhook Attemps:
![https://d.pr/i/UkomqT](https://d.pr/i/UkomqT+)
Full Image: https://d.pr/i/UkomqT

**Comment by billrobbins** [comment]
We had another report in 2971790-zen.  The server time appears to be correct in this case.

![time](https://cdn-std.droplr.net/files/acc_705319/t9Q7Da)
(Link to screenshot:  https://d.pr/i/t9Q7Da)

**Comment by bernattorras** [comment]
Another customer reporting a similar issue with Bancontact: 3079698-zen

**Comment by anant1811** [comment]
Another report in 3067502-zen - Multibanco, and server time is correct.

**Comment by phillipcornwell** [comment]
There is another report of this with Multibanco on 3179522-zen

The server time is correct in this case.

**Comment by anant1811** [comment]
Another report in 3196078-zen

**Comment by grigorijurasov** [comment]
> Another report in 3196078-zen

Asking them to clear the webhook secret field and test again as per @dechov comment. If that works - perhaps we can document it as a workaround somewhere until we are able to figure out why is this happening.

**Comment by iMazed** [comment]
@grigorijurasov I've asked 3179522-zen to do this. Mentioning here for documentation purposes.

**Comment by nixiack** [comment]
Similar issue 32688810-zen.  Asked them to clear webhook secret per previous comment.

**Comment by dougaitken** [comment]
Another example in 1001563-hc

The sources wasn't charged within 6 hours:
> A source with ID src_1HRExxxxxxxxxxxxxxx is chargeable2020/09/14 12:11:26
> A source with ID src_1HRExxxxxxxxxxxxxxx was canceled 2020/09/14 18:11:40

**Comment by dechov** [comment]
Thanks!

As far as I can tell, there are three sorts of scenarios that could possibly be causing an issue and that are worth distinguishing:
- Webhook **not configured**: despite messaging about how one must be configured, it's possible AFAICT that at least some of these cases are due to the webhook not being set up, especially since the payment could still work a (potentially large) portion of the time (i.e. when the source is already chargeable upon redirect). In these cases, there would be _nothing_ under "Webhook attempts" in the Stripe dashboard view for the `source.chargeable` event, even within the 15 days before those logs disappear (unless a webhook is configured for a wrong/different endpoint).
- Webhook **failing**: the webhook was attempted but delivery failed. That could happen for various reasons, but this is the case for which I'd recommended checking on the server time, since that was the only sufficiently plausible reason why validation might fail if a webhook secret is set. There could also be a bug in the validation logic but I wasn't able to find one when inspecting the code.
- Webhook **successful**: if the webhook was successfully delivered but the source was still not charged, that would indicate a bug in the handling logic, though I didn't manage to identify one when inspecting the code.

To clarify, the failure mode described in the issue involves a source being cancelled 6 hours after becoming chargeable. It looks like the screenshots in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1196#issuecomment-643795101 relate to either a non-issue or an issue of a different kind, considering that the chargeable `src_1Golxxxxxxxxxxxxxxx` does not end up cancelled.

Any other information or ideas to help narrow down the cause(s) would be helpful.

**Comment by SiobhyB** [comment]
3370347-zen seems to be another case. I've asked the user to try clearing the webhook field, as per the previous comments and will report back.

**Comment by dechov** [comment]
@SiobhyB Thanks for reporting – I've added a note on the ticket:

> According to https://stripe.com/docs/sources/ideal#sources-expiration, chargeable sources are only canceled after 6 hours, not a few minutes – other than that, it does sound like the issue we are investigating in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1196
>
> If possible, it would be useful to know whether there was a webhook delivery attempt for the `source.chargeable` event, and whether it succeeded (it would be listed on their Stripe dashboard).

**Comment by bernattorras** [comment]
Another customer experiencing this issue here: 3444798-zen

**Comment by tsetend** [comment]
One more customer experiencing this issue with Multibanco: 3493263-zen

**Comment by allendav** [comment]
Recommending this for our next or next next sprint

**Comment by GeoJunkie** [comment]
Another report in 3506008-zen

**Comment by fylgjur** [comment]
Another report here: 3553414-zen. Timer was out by 10 mins but correcting it didn't fix the issue. They removed the webhook secret which resolved it. Please follow up if possible.

**Comment by robertocyrino** [comment]
One more in 3542681-zen. Server time is correct.

**Comment by madeincosmos** [comment]
Another site with possibly the same problem: 3566166-zen

**Comment by v18** [comment]
Observations from the various tickets:

- some merchants did not set up their webhooks URL at all, causing failures (we may want a separate issue for this)
- some merchants had their system time off, which caused faillures
- some merchants saw the error go away after clearing the webhook secret field 

Example of webhook secret field being cleared from 3553414-zen:

> We suggested the workaround of emptying the 'webhook secret' field, and when you did this the error was resolved.
> The webhook secret is an optional field so this should not cause you any problems.

Same in 3196078-zen. That seems like a good place to check for possible bugs related to this.

**Comment by mouligreenlaw** [comment]
Adding this ticket as it describes similar issues using SOFORT
3638259-zen

**Comment by mouligreenlaw** [comment]
Adding this ticket again as it now describes similar issues using EPS via Stripe
3638259-zen

**Comment by dougaitken** [comment]
Another SOFORT example in 27503071-hc

Around 1/3 of orders are having the `This chargeable source timed out after 6 hours and the funds were returned to the customer's account.` message and are canceled after six hours.

They have cleared the webhook secret for now and will monitor the failures. 

As an interesting note, @v18, the developer shared logs and there is a difference in the request - one is a webhook and one is a redirect - unsure if this has any bearing on the cause?

**Comment by mbaierl** [comment]
> They have cleared the webhook secret for now and will monitor the failures.

So far it works perfectly fine - no more reports of Sofort payments not working.

**Comment by dechov** [comment]
> As an interesting note, @v18, the developer shared logs and there is a difference in the request - one is a webhook and one is a redirect - unsure if this has any bearing on the cause?

@dougaitken As I understand it, the reason for the webhook handling here is to capture the charge in the event that the source was not capturable in time for the redirect back to the site (causing the request to fail).

If the webhook always fails or isn't set up, the charges would likely still succeed a portion of the time just from the redirect handling alone. (Not sure if that entirely explains the "intermittent" failures, or if the _webhook_ is working "intermittently" for some merchants.)

**Comment by v18** [comment]
I'm marking this issue as non-critical for now. The reason is that we are not able to reproduce or pinpoint the problem that's happening. Since there are a few different scenarios being reported, it seems there are a few different problems. We've been working and are continuing to work to handle webhooks errors in a better way, to help illuminate the problem.

As new issues come in, could you please add some more details to the post, in addition to the ticket:
- which payment method was being used
- was there a chargeable source? what's the timestamp?
- was the chargeable source later cancelled? what's the timestamp?
- were there any failed webhooks related to the charge?
- does the merchant use a webhook signing secret?
- do charges for that payment method fail repeatedly for the store owner? if so, are there any patterns?

**Comment by mbaierl** [comment]
> * which payment method was being used

SOFORT

> * was there a chargeable source? what's the timestamp?

2021-02-25T11:06:26+00:00 DEBUG 
====Stripe Version: 4.8.0====
====Start Log====
Info: Begin creating SOFORT source
====End Log====

> * was the chargeable source later cancelled? what's the timestamp?

2021-02-25T17:07:20+00:00 DEBUG 
====Stripe Version: 4.8.0====
====Start Log====
Canceled webhook abort: Order was not processed by Stripe: 32631
====End Log====

In Stripe:
- Source was canceled and funds returned to the customer
- This chargeable source timed out after 6 hours and the funds were returned to the customer's account.

Source details
ID:src_XXX
Type:sofort
Flow:redirect
Status:canceled
Usage:single_use
Date:2021/02/25 12:06:26

Events from Stripe:
A source with ID src_XXX was canceled		2021/02/25 18:07:19
A source with ID src_XXX is chargeable		2021/02/25 12:07:11


> * were there any failed webhooks related to the charge?

No, 0% error rate.

> * do charges for that payment method fail repeatedly for the store owner? if so, are there any patterns?

No, sometimes they fail (more often if the Webhook Signing Secret is active), this weekend all of them failed. I can't see any pattern.

**Comment by theabhig** [comment]
I just came across this on 3802714-zen.

**Which payment method was being used** 

Stripe iDEAL

**Was there a chargeable source? What's the timestamp?** 

User didn't specify this.

**Was the chargeable source later cancelled? what's the timestamp?**

User confirmed both of the failed payments were cancelled 6 hours later.

**Were there any failed webhooks related to the charge?**

User didn't specify this.

**Do charges for that payment method fail repeatedly for the store owner? if so, are there any patterns?**

No patterns. User said payments were processed fine in iDEAL both before and after their customer's one and their own test one.

**Comment by orhanfernandes** [comment]
Similar issue on 4132892-zen

- Which payment method was being used

Stripe Multibanco (The source was cancelled after 6 hours)

- what's the timestamp for the chargeable source?

05/07/2021 13:05:45

- when the chargeable source was later cancelled, what's the timestamp?

05/07/2021 19:05:48

- were there any failed webhooks related to the charge?

No failed webhooks reported

- do you use a webhook signing secret?

Yes

- do charges for that payment method fail repeatedly? if so, are there any patterns?

User said " This payment method is new, so I never managed to make it work"

They also mentioned:

"I tried to expire the secret but still not working. I just created a new purchase order and everything goes right until the source changes to “is chargeable”, and then nothing happens… On the woocommerce side the order is still pending payment confirmation."

**Comment by theabhig** [comment]
Similar issue in 4179681-zen

**Which payment method was being used?**

iDEAL

**Was there a chargeable source? what's the timestamp?**

User didn't specify this

**Was the chargeable source later cancelled? What's the timestamp?**

It looks like order was placed at 18:20 on 23 July and was cancelled 2 hours later.

**Were there any failed webhooks related to the charge?**

None reported.

**Does the merchant use a webhook signing secret?**

Unsure.

**Do charges for that payment method fail repeatedly for the store owner? If so, are there any patterns?**

No patterns. So far this seems to be the first (and only) order this has happened on. I am waiting for more information from them and can add this once they reply.

**Comment by neilmccreadie** [comment]
another instance in 4371241-zen

**which payment method was being used**

SOFORT

**was there a chargeable source? what's the timestamp?
was the chargeable source later cancelled? what's the timestamp?**

Order was cancelled exactly 6 hours after initial order. 

**were there any failed webhooks related to the charge?**

Uncertain

**does the merchant use a webhook signing secret?**

Yes, have suggested removing this.

**do charges for that payment method fail repeatedly for the store owner? if so, are there any patterns?**

Seems to be the first occurrence, so no patterns as yet.

**Comment by slash1andy** [comment]
another instance in 4456797-zen

**which payment method was being used**

iDeal

**was there a chargeable source? what's the timestamp?  
was the chargeable source later cancelled? what's the timestamp?**

Order was cancelled exactly 6 hours after initial order.

**were there any failed webhooks related to the charge?**

none reported

**does the merchant use a webhook signing secret?**

Yes, have suggested removing this.

**do charges for that payment method fail repeatedly for the store owner? if so, are there any patterns?**

Seems to be the first 2 occurrences

**Comment by darcie** [comment]
5174763-zen 

**which payment method was being used**
iDeal 

**was there a chargeable source? what's the timestamp?
was the chargeable source later cancelled? what's the timestamp?**

Not all information was provided, but a charge appeared on the customer's bank account and was later cancelled. The transaction was not recorded in Stripe. 

**were there any failed webhooks related to the charge?
does the merchant use a webhook signing secret?**
Unknown 

**do charges for that payment method fail repeatedly for the store owner? if so, are there any patterns?**
First occasion. 

We do know they didn't have their checkout page set up correctly.

**Comment by maxlaf** [comment]
5424675-zen

**which payment method was being used**

iDeal

**was there a chargeable source? what's the timestamp?
was the chargeable source later cancelled? what's the timestamp?**

Order was cancelled 6 hours after the source was started. A charge appeared on the customer's bank account but there's no transaction recorded in Stripe.

**were there any failed webhooks related to the charge?**

`Canceled webhook abort: Order was not processed by Stripe:`

**does the merchant use a webhook signing secret?**

Yes, have suggested removing this.

**do charges for that payment method fail repeatedly for the store owner? if so, are there any patterns?**

Appears to be the first occurrence.

**Comment by thenbrent** [comment]
I'm going to close this as the last report was Jul, 2022 and the penultimate report was May 2022. This could mean the issue has been resolved given no reports in the last 12 months, and just 2 reports in the last 18 months.

We will also be focusing future effort on the UPE based APMs instead of legacy elements integration, which these reports relate too (A8C reference for the roadmap here: pc4etw-Jq-p2)

In cases where the core issue isn't resolved, it is still clear there are reports of the same symptom but for many different causes here (as summarized well by Val's comment https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1196#issuecomment-782562896 but also evident from later reports, where the Checkout was not setup correctly, which could also cause this.) While each individual cause can rightly be considered a critical issue, one GitHub Issue to track so many different problems is not an effective way to resolve those issues, so please create a new issue more specific to the cause rather than the symptom if one doesn't already exist.
